lib/ansible/cli/adhoc.py

```diff
@@ -64,10 +64,18 @@ def post_process_args(self, options):
 
         return options
 
-    def _play_ds(self, pattern, async_val, poll):
-        check_raw = context.CLIARGS['module_name'] in C.MODULE_REQUIRE_ARGS
-
-        mytask = {'action': {'module': context.CLIARGS['module_name'], 'args': parse_kv(context.CLIARGS['module_args'], check_raw=check_raw)},
+    def _play_ds(self, pattern, async_val, poll, loader):
+        if context.CLIARGS['module_args'].startswith(u"@"):
+            # Argument is a YAML file (JSON is a subset of YAML)
+            args = loader.load_from_file(context.CLIARGS['module_args'][1:])
+        elif context.CLIARGS['module_args'][0] in [u'/', u'.']:
+            raise AnsibleOptionsError("Please prepend extra_vars filename '%s' with '@'" % context.CLIARGS['module_args'])
+        elif context.CLIARGS['module_args'][0] in [u'[', u'{']:
+            args = loader.load(context.CLIARGS['module_args'])
+        else:
+            check_raw = context.CLIARGS['module_name'] in C.MODULE_REQUIRE_ARGS
+            args = parse_kv(context.CLIARGS['module_args'], check_raw=check_raw)
```

Comment:
I would reverse order and have parse_kv go first, since it will also still be the most common usage, also the current detection will create many issues, we have used it before (still do for safe_eval).

Use  a try/fail approach on parse_kv to try yaml parsing afterwards and error out if all fail

the module_require_args check should not be limted to kv, but also force skip yaml attempt

Reply:
the try/fail method will not work because the parse_kv at that moment does not fail but it creates raw params (`{'_raw_params': '{"msg":[{"test":123}]}'}`)